### ISSUE TITLE
Add remaining health connect sensors

### ISSUE THREAD

**Issue by dshokouhi** [home-assistant/android#4804]
This request will be used to track the remaining sensors we can add according to: https://developer.android.com/health-and-fitness/guides/health-connect/plan/data-types as I am sure we will get more requests

- [x] Active calories burned 
- [ ] Basal body temperature
- [ ] Basal metabolic rate
- [x] Blood glucose
- [ ] Blood pressure
- [x] Body fat
- [ ] Body temperature
- [ ] Body water mass
- [ ] Bone mass
- [ ] Cervical Mucus
- [ ] Cycling pedaling Cadence
- [x] Distance
- [x] Elevation gained
- [ ] Exercise session
- [x] Floors climbed
- [x] Heart rate
- [ ] Heart rate variability rmmsd
- [ ] Height
- [ ] Hydration
- [ ] Intermenstrual bleeding
- [ ] Lean body mass
- [ ] Menstruation flow
- [ ] Menstruation period
- [ ] Nutrition (will require extra care as its probably lots of sensors in it of itself)
- [ ] Ovulation test
- [ ] Oxygen saturation
- [ ] Planned exercise
- [ ] Power
- [ ] Respiratory rate
- [ ] Resting heart rate
- [ ] Sexual activity
- [ ] Skin temperature
- [x] Sleep session (duration)
- [ ] Speed
- [ ] Steps cadence
- [x] Steps
- [x] Total calories burned
- [ ] VO2 max
- [x] Weight
- [ ] Wheelchair pushes

**Comment by dshokouhi** [comment]
Several of these are in progress in: https://github.com/home-assistant/android/pull/4803

- Basal metabolic rate
- Blood pressure diastolic
- Blood pressure systolic
- Body temperature
- Bone mass
- Height
- Lean body mass
- Oxygen saturation
- Respiratory rate
- VO2 max

**Comment by dshokouhi** [comment]
Exercise session (and related metrics) will need to have discussion around how it works or if it makes sense to add as records will be delayed since the app does not get notified for it. The Wear OS activity state sensor is better fit IMO but only applies for those watches. If we do add it the state as type will require a force update so we can update back to back records of the same type. End time will be unique but may not display well in the frontend without some effort. 

For cycle tracking might be best to hold off until we get a user request so we can discuss what data is best to add

skin temp requires extra consideration in how to report it, do we go by average or latest record?

remaining sensors should also be requested and discussed before we add them

**Comment by FuskaS2** [comment]
I'm glad this is in progress, waiting for menstruation tracking like kid for Christmas.

**Comment by Achi596** [comment]
The latest update with the new metrics was pretty cool. Can't wait for calories consumed to be added.

**Comment by cpeng03d** [comment]
I absolutely need the body temperature sensor exposed in HA!
I'd say use average over the last 10min or so.

**Comment by smmarczak** [comment]
I'm excited about this work! Should I be able to see the glucose readings from health connect in the companion app yet?

**Comment by dshokouhi** [comment]
> I'm excited about this work! Should I be able to see the glucose readings from health connect in the companion app yet?

unfortunately we were only approved for a small set of permissions, we plan to go again for some of the original sensors in sometime to see if we can qualify for it. We are not confident we will be approved despite previous approval.

**Comment by smmarczak** [comment]
Thank you. Really appreciate the work. Buzz kill they are so strict on approving this. Appreciate ya!smm
-------- Original message --------From: Daniel Shokouhi ***@***.***> Date: 1/22/25  10:17 AM  (GMT-06:00) To: home-assistant/android ***@***.***> Cc: Stan Marczak ***@***.***>, Comment ***@***.***> Subject: Re: [home-assistant/android] Add remaining health connect sensors (Issue #4804) 

I'm excited about this work! Should I be able to see the glucose readings from health connect in the companion app yet?

unfortunately we were only approved for a small set of permissions, we plan to go again for some of the original sensors in sometime to see if we can qualify for it. We are not confident we will be approved despite previous approval.

**Comment by jonathan-morton** [comment]
> > I'm excited about this work! Should I be able to see the glucose readings from health connect in the companion app yet?
> 
> unfortunately we were only approved for a small set of permissions, we plan to go again for some of the original sensors in sometime to see if we can qualify for it. We are not confident we will be approved despite previous approval.

Has there been any consideration for a version that includes Google features that are blocked by Google? 

Might make the pipeline more complex, since it would require changing the manifest for play store releases to remove permissions, and ideally would include an alert for a new version to download manually. But HA is in a unique position where the Play Store isn't necessarily a barrier to user acquisition (and that's not even necessarily a strict goal)

 I know there is already a separate FDroid version that is missing location features.

**Comment by ImSkully** [comment]
> [dshokouhi on Nov 21, 2024](https://github.com/home-assistant/android/issues/4804#issuecomment-2492334621)
>  If we do add it the state as type will require a force update so we can update back to back records of the same type. End time will be unique but may not display well in the frontend without some effort.

What is the latest progress on multiple sensor updates to sensors of the same type historically under the same Health Connect entry? Currently only seeing sensors update using the latest value to the entry, though each entry spans several hours and HA misses all prior updates that may have occurred.

**Comment by Dizzyikea** [comment]
Body Water Mass would be great, it's the only one missing from the list I need to monitor post heart attack to improve my vascular health

**Comment by fgsfds** [comment]
Really miss nutrition sensor

**Comment by TimoPtr** [comment]
> Really miss nutrition sensor

We have hard time to pass the Google play validation when adding new sensors. To anyone that wants a new sensor, please write a sample of use case to help us explain to Google how this feature would improve your experience.

**Comment by fgsfds** [comment]
> To anyone that wants a new sensor, please write a sample of use case to help us explain to Google how this feature would improve your experience.

As for me, I want to have a consumed calories sensor on my dashboard so my relatives can bully me if I eat too much.

**Comment by TimoPtr** [comment]
> > To anyone that wants a new sensor, please write a sample of use case to help us explain to Google how this feature would improve your experience.
> 
> As for me, I want to have a consumed calories sensor on my dashboard so my relatives can bully me if I eat too much.

Unfortunately if we write something like that Google is going to reject our app. We need something related to health or fitness.

**Comment by f5ks** [comment]
Something like this would suffice?

Integrating a consumed calories sensor into the dashboard would provide users with a more complete and data-driven view of their overall health and wellness. The sensor would enable accurate tracking of daily caloric intake, allowing for meaningful comparison with energy expenditure data gathered from activity and fitness sensors. Presenting this information cohesively within the interface would facilitate a better understanding of energy balance and promote healthier lifestyle choices.

This capability would support key health objectives such as weight management, dietary optimization, and the development of sustainable nutrition habits. For fitness enthusiasts and individuals following structured health plans, real-time caloric insights could enhance motivation, improve adherence to personal goals, and contribute to more effective long-term outcomes. By providing actionable, evidence-based feedback, the consumed calories sensor would significantly elevate the utility and impact of the platform’s health monitoring features.

-------------------

Yes it is mostly generated, but what else to fight with bureaucracy...

**Comment by TimoPtr** [comment]
> nsumed calories sensor into the dashboard would provide users with a more complete and data-driven view of their overall health and wellness. The sensor would enable accurate tracking of daily caloric intake, allowing for meaningful comparison with energy expenditure data gathered from activity and fitness sensors. Presenting this information c

Something like that yes. For now we have trouble getting the new update out because of these kind of sensors. I hope we are going to be able to get it out. The experience was not fun so we're going to wait before adding new sensors again, we won't add any new health sensors before 2026.

**Comment by ptoews** [comment]
To add another, more specific use case: I want to create a dashboard to plot calorie intake and weight together, so that I can see how my calorie intake influences my weight. For example if the calorie intake I calculated for maintenance actually causes my weight to stay the same (since the calculation is just an approximation and it's hard to factor in different kinds of physical activity). Doing this in Home Assistant would be the best because it gives maximum flexibility so that I can e.g. use the Grafana integration and let it generate trend lines (since weight fluctuates a lot). Overall this would allow me to accurately align my calorie intake to my goals.

**Comment by amykapernick** [comment]
In my case I'd like access to hydration in home assistant so I could create a widget button on my home screen to log each glass of water I drink. The multi step process when logging water means I generally don't keep track of it

**Comment by Tobypoodle** [comment]
Really looking forward to the nutrition data!

**Comment by sabatmonk** [comment]
In my case, I use HA as a substitute for API for some apps (looking at you MacroFactor) so having this data in HA allows me to add it to my health and fitness dashboard, and allows me to have historical data for querying.

**Comment by utxKalEl** [comment]
I wanted to share a real-world usage scenario from our household setup that highlights some limitations in the current Health Connect integration in the Home Assistant Android app.

First of all, thank you for the work that has already been done around Health Connect sensors. The addition of sensors like “Active calories burned” (from #4452) is a great start and already enables some useful dashboards.

However, after using this extensively in a real household setup, I noticed some limitations that make it difficult to use the data correctly.

My setup involves multiple Xiaomi watches syncing into Health Connect, and Home Assistant reading the data through the Android app.

One important issue is how active calories are currently handled.

Right now the “Active calories burned” sensor only reports the latest pushed value instead of aggregating the records. Many fitness apps (like Xiaomi Fitness) push activity data in batches. For example they may sync two 30 minute blocks when the app is opened. This means Home Assistant only sees the latest value and effectively loses part of the data.

This becomes particularly problematic when trying to calculate daily TDEE (total daily energy expenditure).

Currently in Home Assistant the “Total calories burned” sensor appears to represent a combination of BMR and activity. What makes this confusing is that basal metabolic rate data from Google Fitness seems to remain available even after all other Health Connect records have been cleared.

So something is clearly reading BMR data, but it is not exposed as its own sensor. Instead it appears implicitly inside the total calories value.

For advanced use cases it would be extremely helpful if BMR was exposed as a separate sensor rather than implicitly included in total calories.

In our dashboards we calculate things like:
Total calories burned = Basal metabolism + Active calories

Without a dedicated BMR sensor it becomes difficult to build accurate energy expenditure models.

When calculating BMR we rely on the Mifflin-St Jeor equation, which is widely used in nutrition science and coaching.
My wife work as personal trainers and nutrition advisors in addition to other professions.

Whenever body weight changes (either gain or loss), we recalculate BMR and update TDEE accordingly. Some apps do not update this correctly when body weight changes, which can create incorrect calorie expenditure estimates over time.

Because of this it would be extremely helpful if Home Assistant could expose the underlying BMR data directly instead of having it implicitly mixed into the total calories sensor.

Another missing piece is ExerciseSession data.

Example. 
Google Fit already consumes ExerciseSessionRecord from Health Connect and exposes:

- activity type
- start time
- end time
- active calories
- duration

This is extremely useful because it allows distinguishing between structured exercise sessions and passive daily movement.

Fitness trackers are known to overestimate passive movement (for example walking to the kitchen, getting coffee, short errands etc).

In our testing we see roughly:

- around 20% overestimation on my own watch
- around 40% overestimation on my wife's watch

However manually started exercise sessions (running, HIIT etc) are surprisingly close to physiological estimates when compared with MET-based calculations.

Because of this distinction it would be very useful if Home Assistant exposed something like an exercise session sensor containing attributes such as:

- activity_type
- start_time
- end_time
- active_calories
- duration

This would allow dashboards and automations to separate exercise calories from passive activity calories.

Another important category is nutrition.

Currently I import nutrition data using the FatSecret API. Ideally nutrition data would also be accessible through Health Connect so that everything (activity + nutrition) could be unified in one pipeline.

This would allow much better automation and analysis inside Home Assistant, for example:

- dynamic TDEE calculations
- calorie balance tracking
- activity based reminders
- nutrition alerts
- long term health statistics

In my case this is not only used for personal fitness tracking.
We also track activity and nutrition for our children. They are very active in sports and dance, and one of them has autism and struggles with food intake.

Being able to combine activity, calorie expenditure and nutrition intake in a single Home Assistant dashboard makes it much easier to ensure they are getting the correct energy intake and macronutrients for their development.

Home Assistant is uniquely suited for this type of integration because it allows combining health data with automations and dashboards in a way that no typical fitness app currently does.

From a Health Connect perspective it would be extremely useful if the following data types could be supported more completely:

- BasalMetabolicRateRecord
- ExerciseSessionRecord
- ActiveCaloriesBurnedRecord (aggregated properly)
- TotalCaloriesBurnedRecord
- NutritionRecord

Support for aggregated data types would likely also help solve the batch synchronization issue.

Overall I believe Home Assistant could become an excellent hub for personal health analytics if more of the Health Connect data model was exposed. 

Thanks for reading!

**Comment by schklom** [comment]
About the Sleep data, would it be possible to track entire sleep sessions, including stages (c.f. https://developer.android.com/reference/androidx/health/connect/client/records/SleepSessionRecord)?

Thanks for this app and addition, I appreciate it :)

**Comment by jonferreira** [comment]
Can we still have hope for Menstrual and Ovulation data to ever be implemented in HA? Please pretty please?

**Comment by sturgeo** [comment]
+1 for basic nutrition consumption sensors. It would be beneficial to get the protein, carb and fat consumption values.

**Comment by jezzaaa** [comment]
Let me know if I should raise this as a separate issue.

I'm not really sure if this is a bug in the glucose sensor code in the app, or some other reason (eg frequency of update) but my blood glucose data that shows in HA frequently reverts to some mid-level constant. The data in the Health app shows normal values every minute. The data in HA is every 5+ minutes (as expected) with around 10-20% of values being exactly 139mg/dL.

<img width="1099" height="417" alt="Image" src="https://github.com/user-attachments/assets/773a0c2d-c376-4492-8d4b-d29c1943256d" />

A few days ago this was working just fine, but I was using a different type of glucose sensor that gets its data to Health Connect via a the Libre ap). See the difference here, and note that the last sensor value from the Libre app was 139mg/dL:

<img width="1101" height="410" alt="Image" src="https://github.com/user-attachments/assets/35342244-7d7b-41c7-9d91-a97515bcc66a" />

Is there a way to debug the data as it goes into the app, and when it gets sent to HA? I'm leaning towards a problem with the HA-side data collector rather than something in the app. Is there anywhere else I can look?

**Comment by jpelgrom** [comment]
@jezzaaa Please create a new issue, this is unrelated to adding sensors.
